### PR TITLE
CI: Reconfigure CI to cache built binaries

### PR DESCRIPTION
**Pull request by webmiche**
This PR changes caching of MLIR such that it caches the full build folder (because we need the PythonBindings)  and then conditionally recompiles MLIR on a cache miss.

This brings the CI execution time down to under a minute, but there is an issue atm:

It looks like the Python Bindings are not properly installed, even though I added a CI step (called `diag`) that prints the content of the respective folder, which looks in order.

Maybe @math-fehr  or @PapyChacal could take a look? I feel like I might be missing something very obvious, but I can't find it... ([Python Bindings Installation](https://mlir.llvm.org/docs/Bindings/Python/#pre-requisites)).

### CODE REVIEW

**Comment by webmiche** [comment]
Added that. The CI is running. Let's see it in 2h....

**Review by superlopuh — APPROVED** [review]

**Comment by webmiche** [comment]
Doesn't seem to have worked...

**Comment by math-fehr** [comment]
Hmm, that's indeed annoying :/
I'm wondering if we could just scrap that, since anyway we don't want to keep the `mlir-converter`, right?
Otherwise, maybe it's a missing `export PYTHONPATH`?

**Comment by webmiche** [comment]
I think the PythonPath is exported just before the pytest call.

We also need this to check whether we want to execute the tests in lit. I guess we could change the check once #327 is merged, then we won't need the Python Bindings at all...
But that does not actually solve the problem but only sidestep it....
I will look into it more today...

**Comment by webmiche** [comment]
I think I found the issue: There are requirements that need to be installed for the Python bindings to work, in particular something called `pybind`, which allows interoperation between python and C++11. I am now setting up a cache for the requirements file...

**Comment by webmiche** [comment]
> I think I found the issue: There are requirements that need to be installed for the Python bindings to work, in particular something called `pybind`, which allows interoperation between python and C++11. I am now setting up a cache for the requirements file...

By fixing this issue, I got one step further: It now fails one import later :tada:

**Comment by webmiche** [comment]
I now tried to do the compilation with a restored build directory hoping that it would figure out that most stuff is done already, but that does not seem to work (it just rebuilt the project fully).

I have reverted to my best option and will try to get further there...

**Comment by webmiche** [comment]
I am once again stuck. It seems that the outermost part of the bindings are found (`mlir.ir`), but this immediately does a relative import to `._mlir_libs._mlir.ir`, which cannot be resolved. However, I can print the required file and it looks absolutely fine. I am unsure what is happening here and why it cannot be resolved. I will let it cook for some time, maybe I have another idea.

**Review by superlopuh — CHANGES_REQUESTED** [review]
marking as needs change for now

**Comment by webmiche** [comment]
After removing the mlir_converter, this has developed. The CI only needs 2 min now and we don't fail tests anymore, as the bindings don't actually have to work for our test suite now. We still need (a somewhat dysfunctional version of) them though to check which tests to execute. And for that to work, we need to clone MLIR every time. If we find a different check (like is mlir-opt in the path) we could also circumvent that and reduce the CI execution time by about half again.

What do you guys think? I really dislike installing a broken version of the Python Bindings even it it's only for a hack to know which tests to run...

**Comment by webmiche** [comment]
@math-fehr @georgebisbas @superlopuh

**Comment by webmiche** [comment]
shutil seems to be the way. We are down to 1 min.

I will invalidate the caches and retrigger the CI again. It should then also not build the python bindings in the first place, so the cache will be a bit smaller. Plus, this is how we will deploy it eventually, so we should test it :sweat_smile:

File: .github/workflows/ci-mlir.yml
Comment (georgebisbas) [inline .github/workflows/ci-mlir.yml:69]: merge this if with the above?

File: .github/workflows/ci-mlir.yml
Comment (georgebisbas) [inline .github/workflows/ci-mlir.yml:81]: should we split each argument to a separate line?

File: .github/workflows/ci-mlir.yml
Comment (webmiche) [reply .github/workflows/ci-mlir.yml:27]: I think this was due to my formatting settings in vscode. The next commit will fix it.

File: .github/workflows/ci-mlir.yml
Comment (webmiche) [reply .github/workflows/ci-mlir.yml]: Actually, that's not possible. Both are one action and each step can only perform on single action (at least AFAIU).

File: .github/workflows/ci-mlir.yml
Comment (webmiche) [reply .github/workflows/ci-mlir.yml]: Not sure whether I did this correctly. Should we just see after the merge or should I retrigger the CI? What do you prefer?

**Review by superlopuh — APPROVED** [review]

File: .github/workflows/ci-mlir.yml
Comment (webmiche) [reply .github/workflows/ci-mlir.yml]: I retriggered and it seems to work.

**Pull request merged**
